# HW05: Classification and Deep Learning Essentials (due 31st October)

**As part of NEXT WEEK'S homework you will have to provide feedback to two of your classmates' essays on Moodle (by Friday Nov 3rd, at 23:59).** On Moodle, you will be automatically assigned to the two essays you have to provide feedback to on Monday Oct 30th at midnight, in case you want to start ahead.

In this homework, we focus again on a prediction task (as we did in week 3). Before diving into the coding part of the homework, I would like you to reflect on the following problem and how you would approach it.

*Suppose you are a policy advisor to a developing country government that would like to design a social security program to aid individuals with wages lower than \$1000 per month. However, data collection in this country is very hard due to the lack of technology infrastructures, incentives to misreport income, and geographical barriers. Therefore, you have access to some demographic and employment data at the individual level from all the municipalities in addition to geographical and municipality level features (here, you can be creative about which variables you have access to). However, you have access to income data only for a random set of municipalities.*

*How would you decide how to allocate the transfer using the methods you learned for this course? Be very specific on the method and the main variables you would use.*

**There is no right or wrong answer here. This is just a conceptual exercise to make you think about the methods we are learning about in real-life problems.** You don't need to write a lot about this; 100 words or even a scheme about your solution to this task would be enough.

For this proble, we basically need to predict whether people are making under or over $1,000 per month, based on their available characteristics. Or alternatively, we could interpret this problem as figuring out the cost of this program by estimating the total number of the population who is under the wage threshold. Because the available data is some demographic and employment data from a random set of municipalities, we can use this data to predict population income by municipality. By splitting up the available data into a training and test set, we can create an OLS. The depedent variable would be the amount of people under the $1000 limit, and the independent variables would include total population, average employment types (create a scale where low-paying jobs are 1, high paying jobs are 5, etc), and geographic data (proximity to nearest large city). After an appropriate model has been developed, simply use the available data for other municipalities to estimate the cost of the program.

## Coding Exercise

Another area of research that is increasingly employing machine learning is that of medical research (a great example of it is [Mullainathan and Obermeyer, 2021](https://www.nber.org/papers/w26168)). The correct prediction of who may encounter a critical clinical condition is fundamental for the allocation of treatments. Indeed, both treatment availability and doctors' time are not infinite. Therefore, correctly predicting who may be more likely to experience a heart attack or develop cancer is extremely important to help these people and not waste precious resources at the same time.

In the following, the main goal will be to predict the probability of a heart attack using some health indicators described below. **Note that these are fake data created following the pattern from a dataset with real health indicators.**

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
import seaborn as sns
import matplotlib.pyplot as plt

df = pd.read_csv('data/HW05.csv')

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


**Attribute Information**

- age
- sex: 0 = female; 1 = male
- cp: chest pain type (4 values)
- trestbps: resting blood pressure
- chol: serum cholestoral in mg/dl
- fbs: fasting blood sugar > 120 mg/dl
- restecg: resting electrocardiographic results (values 0,1,2)
- thalach: maximum heart rate achieved
- exang: exercise induced angina
- oldpeak: ST depression induced by exercise relative to rest
- slope: the slope of the peak exercise ST segment
- ca: number of major vessels (0-3) colored by flourosopy
- thal: 0 = normal; 1 = fixed defect; 2 = reversable defect
- target: 0= less chance of heart attack 1= more chance of heart attack

In [3]:
df.head()
print(df['target'].value_counts())


target
1    165
0    138
Name: count, dtype: int64


### XGboost

In this part you will build a classifier for the likelihood of having an heart attack using xgboost. You have to train, validate your classifier and print the most meaningful metrics.

In [4]:
import xgboost as xgb
from xgboost import XGBClassifier

X = df.drop(columns=['target'])
y = df['target']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)
X_valid, X_test, y_valid, y_test = train_test_split(X_test, y_test, test_size=0.5)
##TODO train a classifier using early stopping and the logloss evaluation metric 

xgb_class = XGBClassifier(eval_metric='logloss')
xgb_class.fit(X_train, y_train, early_stopping_rounds=12, eval_set=[(X_valid, y_valid)], verbose=True)


[0]	validation_0-logloss:0.61397
[1]	validation_0-logloss:0.57519
[2]	validation_0-logloss:0.56951
[3]	validation_0-logloss:0.52892
[4]	validation_0-logloss:0.52756
[5]	validation_0-logloss:0.49205
[6]	validation_0-logloss:0.46552
[7]	validation_0-logloss:0.48584
[8]	validation_0-logloss:0.46345
[9]	validation_0-logloss:0.48539
[10]	validation_0-logloss:0.47788
[11]	validation_0-logloss:0.46275
[12]	validation_0-logloss:0.45129
[13]	validation_0-logloss:0.46861
[14]	validation_0-logloss:0.46457
[15]	validation_0-logloss:0.48282
[16]	validation_0-logloss:0.47281
[17]	validation_0-logloss:0.46428
[18]	validation_0-logloss:0.46974
[19]	validation_0-logloss:0.47036
[20]	validation_0-logloss:0.47840
[21]	validation_0-logloss:0.49547
[22]	validation_0-logloss:0.50798
[23]	validation_0-logloss:0.50621


/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/xgboost/sklearn.py:885: UserWarning: `early_stopping_rounds` in `fit` method is deprecated for better compatibility with scikit-learn, use `early_stopping_rounds` in constructor or`set_params` instead.
  warnings.warn(


XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric='logloss',
              feature_types=None, gamma=None, grow_policy=None,
              importance_type=None, interaction_constraints=None,
              learning_rate=None, max_bin=None, max_cat_threshold=None,
              max_cat_to_onehot=None, max_delta_step=None, max_depth=None,
              max_leaves=None, min_child_weight=None, missing=nan,
              monotone_constraints=None, multi_strategy=None, n_estimators=None,
              n_jobs=None, num_parallel_tree=None, random_state=None, ...)

In [5]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
y_pred = xgb_class.predict(X_test)

metrics = confusion_matrix(y_test,y_pred)
print(metrics)
print("accuracy:",accuracy_score(y_test, y_pred))
##TODO plot the confusion metrics and calculate the accuracy score


[[ 8  6]
 [ 1 16]]
accuracy: 0.7741935483870968


**What can you say about the performance of your classifier based on these metrics?**

The performance of the classifier is ok. The .77 accuracy score means that 77% of the predictions are correct. The 4 false positives and 3 false negatives mean that some patients are not given the correct diagnosis. The model has a precision of (9/(9+4)) = .69, and recall of (9/(9+3)) = .75, so 25% of people with high risk are missed, which isn't great. 

**Which metrics are better suited to evaluate this model? Calculate and visualize these metrics, and comment on the performance of the model.**

Because this is a medical model, the most important metric is the recall, because patients who are 'missed' are then at a high risk and don't know it. The f1 score is also important because the precision and recall are both important in any model. The balanced accuracy score is not important because as shown above, there are roughly equal 1 and 0 targets. auc and roc_auc score are generally important in models to show how effective they are. 

In [6]:
from sklearn.metrics import (precision_score, recall_score, f1_score, precision_recall_curve,
                             balanced_accuracy_score, roc_curve, auc, roc_auc_score)

print("Precision score:",precision_score(y_test,y_pred))
print("Recall score:",recall_score(y_test,y_pred))
print("f1:",f1_score(y_test,y_pred))
print("Balanced accuracy score:",balanced_accuracy_score(y_test,y_pred))
print("roc_auc_score",roc_auc_score(y_test, y_pred))
fpr, tpr, unused = roc_curve(y_test,y_pred)
print("auc",auc(fpr,tpr))
precision, recall, unused = precision_recall_curve(y_test, y_pred)
print("prec, rec:",precision,recall)





Precision score: 0.7272727272727273
Recall score: 0.9411764705882353
f1: 0.8205128205128205
Balanced accuracy score: 0.7563025210084033
roc_auc_score 0.7563025210084032
auc 0.7563025210084032
prec, rec: [0.5483871  0.72727273 1.        ] [1.         0.94117647 0.        ]


Based on the output of these metrics, the roc_auc_score, which measures the area under the roc curve, is .75. This is on a scale from 1 to 0, and 1 is perfect, so the score is ok. The recall is .94, which is ok for this scenario because it deals with preventative medicine. Only 6% of high risk patients will be misidentified. Based on the f1 score of .82, the precision and recall together are high, but not great for a high-pressure scenario like this medical context. 

## Deep Learning

Now, you will build an analogous classifier, i.e., with the same objective as the one in the previous part, using a neural network structure. 

In [7]:
import torch
import torch.nn as nn
import torch.optim as optim
## build a MLP (multilayer perceptron) model to predict the outcome using 
# the same predictors as in the XGBoost model.
# the MLP model should have at least 2 hidden layers, ReLU activation

#TODO

nn_model = nn.Sequential(
    nn.Linear(13, 6),  # 1st hidden layer with ReLU activation
    nn.ReLU(),
    nn.Linear(6, 7),  # 2nd hidden layer with ReLU activation
    nn.ReLU(),
    nn.Linear(7, 1)  # Output layer
)

In [8]:
# define the loss function and the optimizer
loss = nn.BCELoss()
learning_rate = 0.01  # You can adjust the learning rate based on your problem
optimizer = optim.SGD(nn_model.parameters(), lr=learning_rate)

In [9]:
#fit the model on the training data
from torch.utils.data import DataLoader, TensorDataset


X_tensor = torch.tensor(X_train.values, dtype=torch.float32)
y_tensor = torch.tensor(y_train.values, dtype=torch.float32)

train_tensor = TensorDataset(X_tensor, y_tensor)

batch_size = 32

train_dataloader = DataLoader(train_tensor, batch_size=batch_size, shuffle=True)


n_epochs = 10
for epoch in range(n_epochs):
    nn_model.train()
    totalLoss = 0

    for vals, outcome in train_dataloader:
        optimizer.zero_grad()  
        
        output = torch.sigmoid(nn_model(vals))
        loss_amt = loss(output, outcome.view(-1, 1))  # Reshape outcome to match output shape
        totalLoss += loss_amt
        
        loss_amt.backward()
        
        optimizer.step()

In [35]:
#compute the test set accuracy, as well as the metrics you picked to evaluate the xgboost model
nn_model.eval()

corr = 0
tot = 0
true_positive = 0
false_positive = 0
false_negative = 0
true_negative = 0

y = [] 
y_pred = []

X_tensor_test = torch.tensor(X_test.values, dtype=torch.float32)
y_tensor_test = torch.tensor(y_test.values, dtype=torch.float32)

test_tensor = TensorDataset(X_tensor_test, y_tensor_test)

with torch.no_grad():
    for vals, label in test_tensor:
        outcome = nn_model(vals)

        pred = (outcome >= .5).float().item()
        predicted = pred == label.item()

        tot += 1
        y.append(label.item())
        y_pred.append(pred)

        if label.item() == 0:
            if pred == 0:
                true_negative += 1
                corr += 1
            else:
                false_positive += 1
        else:
            if pred == 0:
                false_negative += 1
            else:
                corr += 1
                true_positive += 1

print("accuracy:",(corr / tot)*100,"%")

print("precision:",true_positive / (true_positive + false_positive))
print("recall:",true_positive / (true_positive + false_negative))
print("balanced accuracy:",(recall + true_negative / (true_negative + false_positive)) / 2)
print("AUC:",roc_auc_score(y, y_pred))





accuracy: 48.38709677419355 %
precision: 0.5555555555555556
recall: 0.29411764705882354
balanced accuracy: 0.631336405529954
AUC: 0.5042016806722689


**How does the NN-based classifier performs with respect to the XGBOOST one?**

The XGBOOST classifier performs better than the NN. The XGBOOST model has a precision and recall of .72 and .94, which are ok in the context of this problem because while nearly 30% of healthy patients will be told that they have a problem, only 6% of at-risk patients are skipped. For the NN, the precision and recall both take a signficant hit; .55 and .29,respectively. This means that upwards of 70% of at-risk patients are told that they don't have a problem, which is much worse than a coin flip. The accuracy is 48.38%, which means that the classification overall is not helpful. Additionally, even though the metric isn't particularly meaningful for this problem due to the lbel distribution being even, the balanced accuracy score of .63 is not good. The AUC score of .5 also shows that this model is not effective at predicting. 

**OPTIONAL QUESTION: Suppose that one of the classifiers you built performs very well, would you suggest using it to decide on the allocation of treatments to prevent heart attacks?**

No. I would not because the recall is still not tuned highly enough. I would want an extremely high recall for a situation like this, where it is really important to not miss any at-risk patients.